tests/test_umbral.py

```diff
@@ -58,6 +58,120 @@ def test_m_of_n(N, threshold):
     assert sym_key_2 == sym_key
 
 
+def test_kfrag_serialization():
+    pre = umbral.PRE(umbral.UmbralParameters())
+
+    priv_key = pre.gen_priv()
+    pub_key = pre.priv2pub(priv_key)
+
+    frags, _ = pre.split_rekey(priv_key, pub_key, 1, 2)
+    frag_bytes = frags[0].to_bytes()
+
+    assert len(frag_bytes) == 194
+
+    new_frag = umbral.KFrag.from_bytes(frag_bytes,
+                                       umbral.UmbralParameters().curve)
+    assert new_frag.bn_id == frags[0].bn_id
+    assert new_frag.point_key == frags[0].point_key
+    assert new_frag.point_eph_ni == frags[0].point_eph_ni
+    assert new_frag.point_commitment == frags[0].point_commitment
+    assert new_frag.bn_sig1 == frags[0].bn_sig1
+    assert new_frag.bn_sig2 == frags[0].bn_sig2
+
+
+def test_cfrag_serialization():
+    pre = umbral.PRE(umbral.UmbralParameters())
+
+    priv_key = pre.gen_priv()
+    pub_key = pre.priv2pub(priv_key)
+
+    _, capsule = pre.encapsulate(pub_key)
+    kfrags, _ = pre.split_rekey(priv_key, pub_key, 1, 2)
+
+    cfrag = pre.reencrypt(kfrags[0], capsule)
+    cfrag_bytes = cfrag.to_bytes()
+
+    assert len(cfrag_bytes) == 131
+
+    new_cfrag = umbral.CapsuleFrag.from_bytes(cfrag_bytes,
+                                              umbral.UmbralParameters().curve)
+    assert new_cfrag.e1 == cfrag.e1
+    assert new_cfrag.v1 == cfrag.v1
+    assert new_cfrag.bn_kfrag_id == cfrag.bn_kfrag_id
+    assert new_cfrag.point_eph_ni == cfrag.point_eph_ni
+
+
+def test_capsule_serialization():
+    pre = umbral.PRE(umbral.UmbralParameters())
+
+    priv_key = pre.gen_priv()
+    pub_key = pre.priv2pub(priv_key)
+
+    _, capsule = pre.encapsulate(pub_key)
+    capsule_bytes = capsule.to_bytes()
+
+    assert len(capsule_bytes) == 98
+
+    new_capsule = umbral.Capsule.from_bytes(capsule_bytes,
+                                            umbral.UmbralParameters().curve)
+    assert new_capsule.point_eph_e == capsule.point_eph_e
+    assert new_capsule.point_eph_v == capsule.point_eph_v
+    assert new_capsule.bn_sig == capsule.bn_sig
+
+
+def test_reconstructed_capsule_serialization():
+    pre = umbral.PRE(umbral.UmbralParameters())
+
+    priv_key = pre.gen_priv()
+    pub_key = pre.priv2pub(priv_key)
+
+    _, capsule = pre.encapsulate(pub_key)
+    kfrags, _ = pre.split_rekey(priv_key, pub_key, 1, 2)
+
+    cfrag = pre.reencrypt(kfrags[0], capsule)
+
+    capsule.attach_cfrag(cfrag)
+
+    rec_capsule = capsule.reconstruct()
```

Comment:
So, here, I think I prefer a mutation rather than an instantiation.  It gets a little dizzying to have two objects after we refactored them to share a life cycle.  But that can come in a later PR.